### ISSUE TITLE
Enhancement: [no-non-null-assertion] a setting to forbid no-null-assertion to class properties

### ISSUE THREAD

**Issue by mnaoumov** [typescript-eslint/typescript-eslint#11665]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-non-null-assertion/

### Description

I propose to add an option that will not allow non-null-assertion for class properties.

### Fail

```TypeScript
class Foo {
  bar!: string;
}
```

### Pass

```TypeScript
class Foo {
  bar: string = 'baz';
}
```

**Comment by kirkwaiblinger** [comment]
Hi @mnaoumov !

As you note, looks like this can be achieved currently with no-restricted-syntax, selector `[definite=true]` ([playground](https://typescript-eslint.io/play/#ts=5.9.2&showAST=es&fileType=.ts&code=MYGwhgzhAEBiD29oG8BQ1oBMCmAzAlgHb4Au2AglPgOaEC22hJAhAFzQQkBOR1A3OmiF4XOmBCUINeoxLtOPQv1QBfVOpDYSWPEVIUqtBkzYduvAZu3DR4ydONyzi5ahtiJhmSYB0JeADK5koAFACUfNAA9FHQ1FxgAJ7QYFzYYOqoOKCp2NDA8IScQiIeACqJAA4GENhcJPiF7IQAriAg0AA%2BzhZupeIV1ZJ1DYXMfEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaaFAF2gEsBjUxAE0OQE9dSBDAD3TAG1IB3NtFyQANJnDYcyREhr5o3SDzqIAZhVwVaAXnKxEAXUiSIAX0OmwAemthxkAAKkmABxRVKr0oRTxNpNYEhAS4IQjwhGzIMtCkFPgiGAJCIpJmIGZAA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)).

Also I found https://github.com/typescript-eslint/typescript-eslint/issues/9782, where this was previously declined.

In addition to the reasoning there, I think another case that makes this difficult to motivate as a general purpose rule is that there's not really an alternative to a definite assignment assertion in very legitimate cases. Whereas the non-null assertion can be replaced with `assert(x != null)`, cases where a definite-assignment assertion is required don't allow the value even to be _read_.  For instance,

[Playground](https://typescript-eslint.io/play/#ts=5.9.2&showAST=es&fileType=.ts&code=GYVwdgxgLglg9mABABwE5wLYwM4FMDqMUAFgEq7ZwA2AbrqgBQCUiA3gFCKJW5SKoVqdAAroseAFyJmiALwA%2BRDTgwAJgG5OiCAmx80mHLjmIwuAO6JRhvAB5la%2BQwYDKtXCwVstXV0NzW4say-ILumlwAvkyaWgD0cYgAKqgAnojgOhgYuGCwYADmiCQ42nCquACE8YkAmnAg2gCGYADkfLh0SAD6UBRQ3YjAcKimIFRUOMRm2NiIxPS4NTWIMMDSfu6BRnIhYONULBxcCVzFxOiWAKKo6IwARCVz2MQNVKptfABGxshwszAvjxKvcYitIuwtAIoCBUEhjmdThAWogEFR0ps6OdSvQ7ohzERiIgmqYEABafYTYmzeiwBDzRaIHyIU4jYmICrAGBgIi4MlNAEFMA5PLUvCoOlIJp8IhzCoQKhNVDS%2BBgZmY3BSDXbPAAGkhZxQYiMWkimkiQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaaFAF2gEsBjUxAE0OQE9dSBDAD3TAG1IB3NtFyQANJnDYcyREhr5o3SDzqIAZhVwVaAXnKxEAXUiSIAX0OmwAemthxkAAKkmABxRVKr0oRTxNpNYEhAS4IQjwhGzIMtCkFPgiGAJCIpJmIGZAA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

### Code

```ts
function promiseWithResolver() {
  let resolvePromise: () => void;
  const promise = new Promise<void>((resolve) => {
    resolvePromise = resolve;
  });

  // Try uncommenting this code!
  // You can't even _test_ for nullishness here
  //
  // if (resolvePromise == null) {
  //   throw Error("this shouldn't be possible!");
  // }

  return {
    // can only resolve this error with a non-null assertion here 
    // or a definite-assignment assertion at its declaration
    resolve: resolvePromise,

    promise
  };
}
```

<details>
<summary>Expand</summary>

### Eslint config

```json
{
  "rules": {
    "no-restricted-syntax": ["warn", {
      "selector": "[definite=true]"
    }]
    // , "@typescript-eslint/no-non-null-assertion": "warn"
  }
}
```

</details>

### TypeScript config

```json
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `8.45.0` |
| `@typescript-eslint/parser` | `8.45.0` |
| `TypeScript` | `5.9.2` |
| `ESLint` | `9.36.0` |
| `node` | `web` |

---

So, with that in mind, I'm -1 to reconsidering this proposal.

**Comment by mnaoumov** [comment]
@kirkwaiblinger 

Thanks for the feedback

I think your example can be rewritten as

```ts
function promiseWithResolver() {
  let resolvePromise: (() => void) | undefined = undefined;
  const promise = new Promise<void>((resolve) => {
    resolvePromise = resolve;
  });

  if (!resolvePromise) {
     throw Error("this shouldn't be possible!");
  }

  return {
    resolve: resolvePromise,
    promise
  };
}
```

while what I requested could be rewritten as

```ts
class Foo {
  _bar?: string;

  get bar(): string {
    if (this._bar === undefined) {
      throw new Error('bar is not initialized');
    }

    return this._bar;
  }

  set bar(value: string): void {
    this._bar = value;
  }
}
```

**Comment by mnaoumov** [comment]
My biggest concern is that with the approach I suggest, we will definitely get a runtime error in-place. With the `!` hack we might get `undefined` which we will pass further and it might fail in a different unpredictable place.

**Comment by kirkwaiblinger** [comment]
@mnaoumov True! In theory you could do this... It's rather extreme IMO for the `resolvePromise` case where we can see with our eyes that the condition is impossible. 

In the case of `Foo#bar`, it has the drawback of turning the class from (nearly) a POJO to something with getters and such (which has its own problems, since you can't conditionally initialize the value using an idiomatic `undefined` check, like `foo.bar ??= 'value'`). That's maybe more of a valid choice if your application really requires it, but again, it's hard to justify it being generally applicable. 

My argument here is really that by providing an option for this, we implicitly endorse its use, which may end up steering non-expert users towards _worse_, not better code.

**Comment by mnaoumov** [comment]
@kirkwaiblinger 

Again, thanks for the feedback. I agree that it's hard sometimes to write code that is both elegant and error-prone.

Not much to discuss here. I didn't find the #9782 because I didn't use the right terms to look for, and it was already rejected. So I think this is going to be closed as well.

So I want to be crystal clear. I am not defending this particular FR, I just trying to have an interesting discussion.

With the previous approach, your example `foo.bar ??= 'value'` would be flagged by another rule: https://typescript-eslint.io/rules/no-unnecessary-condition/

And one last example that it is not possible to achieve total purity anyways.

```ts
let flag = false;
const lambda = () => { flag = Math.random() > 0.5;  };
lambda();
if (flag) { // will be flagged by the rule I mentioned above
}
```

**Comment by JoshuaKGoldberg** [comment]
In terms of the feature request, I agree with Kirk's points. It would be great if we could add lint rules for every single "escape hatch" in the type system. But, alas, the reasons Kirk stated are why we cannot in this case. 

To anybody interested in making a dedicated rule/plugin, https://typescript-eslint.io/developers has docs:

* https://typescript-eslint.io/developers/custom-rules: for making the rule itself
* https://typescript-eslint.io/developers/eslint-plugins: for packaging it as a plugin

N.b. I think this would be best as a separate rule from no-non-null-assertion. It's technically a _definite assignment assertion_.

Cheers!
